scan-payment/src/main/java/com/getbouncer/scan/payment/ml/ExpiryDetect.kt

```diff
@@ -0,0 +1,136 @@
+package com.getbouncer.scan.payment.ml
+
+import android.content.Context
+import android.graphics.Bitmap
+import android.util.Size
+import com.getbouncer.scan.framework.Loader
+import com.getbouncer.scan.framework.ResourceLoader
+import com.getbouncer.scan.framework.ml.TFLAnalyzerFactory
+import com.getbouncer.scan.framework.ml.TensorFlowLiteAnalyzer
+import com.getbouncer.scan.framework.util.indexOfMax
+import com.getbouncer.scan.payment.R
+import com.getbouncer.scan.payment.hasOpenGl31
+import com.getbouncer.scan.payment.scale
+import com.getbouncer.scan.payment.toRGBByteBuffer
+import org.tensorflow.lite.Interpreter
+import java.io.FileNotFoundException
+import java.nio.ByteBuffer
+
+private val TRAINED_IMAGE_SIZE = Size(80, 36)
+
+private const val NUM_CLASS = 11
+private const val NUM_PREDICTIONS = 17
+private const val BACKGROUND_CLASS = 10
+
+val ASPECT_RATIO = TRAINED_IMAGE_SIZE.height.toFloat() / TRAINED_IMAGE_SIZE.width.toFloat()
+
+class ExpiryDetect private constructor(interpreter: Interpreter) :
+    TensorFlowLiteAnalyzer<ExpiryDetect.Input, ByteBuffer,
+        ExpiryDetect.Prediction,
+        Array<Array<Array<FloatArray>>>>(interpreter) {
+
+    data class Input(val image: Bitmap)
+
+    data class Prediction(val expiry: Expiry?)
+
+    data class Expiry(val month: Int, val year: Int) {
+        override fun toString() = "${month.toString().padStart(2, '0')}/$year"
+    }
+
+    data class Digit(val digit: Int, val confidence: Float)
+
+    override val name: String = Factory.NAME
+
+    override fun buildEmptyMLOutput() = arrayOf(
+        arrayOf(
+            Array(17) {
+                FloatArray(NUM_CLASS)
+            }
+        )
+    )
+
+    override fun interpretMLOutput(data: Input, mlOutput: Array<Array<Array<FloatArray>>>): Prediction {
+        val output = mutableListOf<Digit>()
+        for (idx in 0 until 17) {
+            val maxIdx = mlOutput[0][0][idx].indexOfMax()!!
+            output.add(Digit(maxIdx, mlOutput[0][0][idx][maxIdx]))
+        }
+
+        val digits = greedyNonMaxSuppression(output).filter { it != 10 }
+
+        return if (digits.size == 4 || (digits.size == 5 && digits[2] == 1)) {
+            // process if we get exactly 4 digits, OR it's five digits and the middle prediction
+            // is a 1 - this is because we sometimes mistake '/' for '1'
+            val month = digits[0] * 10 + digits[1]
+            val year = 2000 + digits[digits.size - 2] * 10 + digits[digits.size - 1]
+            if (month > 12) {
+                Prediction(null)
+            } else {
+                Prediction(Expiry(month, year))
+            }
+        } else {
+            Prediction(null)
+        }
```

Comment:
Take a look at `CardExpiry.kt` and see if we can leverage some of the methods there to make this cleaner. Those methods were originally designed to deal with strings, but could probably be converted to take individual digits pretty easily.

Reply:
oh nice! didn't realize that existed already